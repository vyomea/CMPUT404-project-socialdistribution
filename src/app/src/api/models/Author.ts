export default interface Author {
  type: "author";
  id: string;
  displayName: string;
  github: string;
  profileImage: string;
  isAdmin: boolean;
}

export type AuthorResponse = Omit<Author, "id"> & {
  id: string; // URL to the author
};

export const authorFromResponse = (data: AuthorResponse): Author => {
  const match = /\/authors\/([^/]+)\/?$/.exec(data.id);
  if (match === null) {
    throw new Error(`Invalid author URL ${data.id}`);
  }
  return {
    ...data,
    id: match[1],
  };
};

export type AuthorRequest = Omit<Author, "id"> & {
  id: string; // URL to the author
};

export const authorToRequest = (
  author: Author,
  baseUrl: string
): AuthorRequest => {
  return {
    ...author,
    id: `${baseUrl}/authors/${author.id}`,
  };
};
