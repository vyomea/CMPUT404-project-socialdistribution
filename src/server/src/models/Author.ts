import { BelongsTo, DataTypes, HasMany, Model } from 'sequelize';
import db from '../db';
import { v4 as uuidv4 } from 'uuid';
import Post from './Post';
import Comment from './Comment';
import Follower from './Follower';
import Node from './Node';
import CommentLike from './CommentLike';
import PostLike from './PostLike';
import FollowRequest from './FollowRequest';

class Author extends Model {
  declare id: typeof uuidv4;
  declare email: string;
  declare passwordHash: string;
  declare displayName: string;
  declare github: string;
  declare profileImage: string;
  declare isAdmin: boolean;
  declare verified: boolean;
  static Posts: HasMany;
  static Comments: HasMany;
  static Followers: HasMany;
  static Node: BelongsTo;
  static CommentLikes: HasMany;
  static PostLikes: HasMany;
  declare followers: Follower[];
  declare posts: Post[];
  declare outgoingFollowRequests: FollowRequest[];
  declare incomingFollowRequests: FollowRequest[];
  declare node: Node;
  declare nodeServiceUrl: string;
  declare commentLikes: CommentLike[];
  declare postLikes: PostLike[];
  declare addComment: (comment: Comment) => Promise<void>;
  declare addPost: (post: Post) => Promise<void>;
  declare addFollower: (author: Author) => Promise<void>;
  declare hasFollower: (author: Author) => Promise<boolean>;
  declare removeFollower: (author: Author) => Promise<void>;
  declare getNode: () => Promise<Node>;
}

Author.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      allowNull: false,
      primaryKey: true,
    },
    email: {
      type: DataTypes.STRING,
      allowNull: true,
      unique: true,
    },
    passwordHash: {
      type: DataTypes.STRING,
      allowNull: true,
    },
    displayName: {
      type: DataTypes.STRING,
      allowNull: true,
    },
    github: {
      type: DataTypes.STRING,
      allowNull: true,
    },
    profileImage: {
      type: DataTypes.STRING,
      allowNull: true,
    },
    isAdmin: {
      type: DataTypes.BOOLEAN,
      allowNull: true,
      defaultValue: false,
    },
    verified: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: false,
    },
  },
  {
    sequelize: db,
    modelName: 'Author',
    underscored: true,
  }
);

Author.Posts = Author.hasMany(Post, { onDelete: 'cascade', hooks: true });
Post.Author = Post.belongsTo(Author, { as: 'author' });

Author.Followers = Author.belongsToMany(Author, {
  through: 'followers',
  as: 'follower',
});
Follower.Follower = Follower.belongsTo(Author, {
  as: 'follower',
  foreignKey: 'followerId',
});
Follower.Author = Follower.belongsTo(Author, {
  as: 'author',
  foreignKey: 'authorId',
});

Author.hasMany(FollowRequest, { as: 'outgoingFollowRequests' });
Author.hasMany(FollowRequest, { as: 'incomingFollowRequests' });
FollowRequest.belongsTo(Author, { as: 'requestee' });
FollowRequest.belongsTo(Author, { as: 'requester' });

Author.Comments = Author.hasMany(Comment);
Comment.Author = Comment.belongsTo(Author, { as: 'author' });

Author.Node = Author.belongsTo(Node, { as: 'node' });
Node.Authors = Node.hasMany(Author, { as: 'authors' });

Author.PostLikes = Author.hasMany(PostLike, { as: 'postLikes' });
PostLike.Author = PostLike.belongsTo(Author, { as: 'author' });
Author.CommentLikes = Author.hasMany(CommentLike, { as: 'commentLikes' });
CommentLike.Author = CommentLike.belongsTo(Author, { as: 'author' });

export default Author;
