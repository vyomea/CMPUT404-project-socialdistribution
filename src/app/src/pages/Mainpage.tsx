import NavBar from '../components/NavBar';
import UserPost from '../components/UserPost';
import Github from '../components/Github';
import styled from 'styled-components';
import Author from '../api/models/Author';
import Post from '../api/models/Post';
import api from '../api/api';
import Add from '../components/Add';
import { useState, useEffect } from 'react';
import Backdrop from '@mui/material/Backdrop';
import { CloseRounded } from '@mui/icons-material';
import Fab from '@mui/material/Fab';
import AddIcon from '@mui/icons-material/Add';
import { List } from '@mui/material';
import { useNavigate } from 'react-router-dom';

// This is for all the stuff in the Main Page
const MainPageContainer = styled.div`
  display: flex;
  flex-direction: column;
  height: ${window.innerHeight}px;
  width: ${window.innerWidth}px;
`;

// This is for the NavBar
const NavBarContainer = styled.div`
  margin-bottom: 5%;
`;

// This is for the posts and New Post Button and Github Activity
const MainPageContentContainer = styled.div`
  display: flex;
  flex-direction: row;
  width: 100%;
`;

const GitContainer = styled.div`
  margin-right: 1%;
`;

interface Props {
  currentUser?: Author;
}

export default function Mainpage({ currentUser }: Props) {
  // For now, mainpage just shows your own posts
  const [posts, setPosts] = useState<Post[] | undefined>(undefined);
  const [open, setOpen] = useState(false);
  const [postsChanged, setpostsChanged] = useState(false);
  const navigate = useNavigate();

  const items2 = [
    {
      Text: 'Home',
      handleClick: () => {
        navigate('/');
      },
    },
    {
      Text: 'Logout',
      handleClick: () => {
        api.logout();
        window?.location?.reload();
      },
    },
  ];

  const handleClose = () => {
    setOpen(false);
  };

  const handleToggle = () => {
    setOpen(!open);
  };

  const handlePostsChanged = () => {
    setpostsChanged(!postsChanged);
  };
  const mockPost = {
    id: '5',
    title: 'title',
    source: 'www.google.com',
    origin: 'www.lol.com',
    description: 'description',
    contentType: 'text/markdown',
    content: '# Lmao',
    image: undefined,
    categories: 'JSON.stringify(category)',
    count: 5,
    published: new Date(),
    visibility: 'visibility',
    unlisted: 'unlisted',
  };
  useEffect(() => {
    api.authors
      .withId("" + currentUser?.id)
      .posts.list(1, 10)
      .then((data) => setPosts(data));
  }, [currentUser?.id, postsChanged]);

  return (
    <MainPageContainer>
      <NavBarContainer>
        <NavBar items={items2} />
      </NavBarContainer>
      <Fab
        color="primary"
        aria-label="check"
        style={{
          margin: 0,
          top: 'auto',
          right: 20,
          bottom: 20,
          left: 'auto',
          position: 'fixed',
        }}
        sx={{ color: 'black', background: '#46ECA6', '&:hover': { background: '#18E78F' } }}
      >
        <AddIcon onClick={handleToggle} />
      </Fab>
      {open ? (
        <Backdrop
          sx={{
            color: '#fff',
            zIndex: (theme) => theme.zIndex.drawer + 1,
            width: '100%',
            height: '100%',
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
          }}
          open={open}
        >
          <CloseRounded
            onClick={handleClose}
            sx={{
              '&:hover': {
                cursor: 'pointer',
              },
              marginBottom: '10px',
              borderRadius: '100%',
              border: '1px solid white',
            }}
          />
          <Add currentUser={currentUser} handlePostsChanged={handlePostsChanged} />
        </Backdrop>
      ) : (
        <MainPageContentContainer>
          <List style={{ maxHeight: '100%', overflow: 'auto' }} sx={{ width: '70%', ml: 5 }}>
            <UserPost
              post={mockPost}
              currentUser={currentUser}
              postAuthor={currentUser}
              likes={0}
              handlePostsChanged={handlePostsChanged}
              key={mockPost.id}
            />
            {/* {posts?.map((post) => (
              <UserPost
                post={post}
                currentUser={currentUser}
                postAuthor={currentUser}
                likes={0}
                handlePostsChanged={handlePostsChanged}
                key={post.id}
              />
            ))} */}
          </List>
          <GitContainer>
            <Github username={currentUser?.displayName} />
          </GitContainer>
        </MainPageContentContainer>
      )}
    </MainPageContainer>
  );
}
