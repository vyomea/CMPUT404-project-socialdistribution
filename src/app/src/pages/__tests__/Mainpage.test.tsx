import React from "react";
import { unmountComponentAtNode } from "react-dom";
import renderer from "react-test-renderer";
import Author from "../../api/models/Author";

import Mainpage from "../Mainpage";

let container: any = null;
const testUser: Author = {
  type: "author",
  id: "",
  displayName: "",
  github: " ",
  profileImage: " ",
  isAdmin: false,
};
//issues with container type and its methods

beforeEach(() => {
  // setup a DOM element as a render target
  container = document?.createElement("div");
  document?.body?.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container?.remove();
  container = null;
});

describe("Mainpage suite", () => {
  it("Renders correctly", () => {
    const tree = renderer.create(<Mainpage currentUser={testUser} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
