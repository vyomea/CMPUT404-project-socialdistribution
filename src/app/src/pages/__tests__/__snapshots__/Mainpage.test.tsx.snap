// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mainpage suite Renders correctly 1`] = `
<div
  className="sc-iwjdpV OHnYU"
>
  <div
    className="sc-cxpSdN TLJOs"
  >
    <div
      className="sc-bdvvtL dVmOGZ"
    >
      <div
        className="sc-gsDKAQ dokASy"
      >
        <button
          className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-79h83g-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Logout
        </button>
      </div>
    </div>
  </div>
  <button
    aria-label="check"
    className="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary css-1ckp4x4-MuiButtonBase-root-MuiFab-root"
    disabled={false}
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "bottom": 20,
        "left": "auto",
        "margin": 0,
        "position": "fixed",
        "right": 20,
        "top": "auto",
      }
    }
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="AddIcon"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
      />
    </svg>
  </button>
  <div
    className="sc-llYSUQ gRojfe"
  >
    <ul
      className="MuiList-root MuiList-padding css-xgw77h-MuiList-root"
      style={
        Object {
          "maxHeight": "100%",
          "overflow": "auto",
        }
      }
    />
    <div
      className="sc-iJKOTD kGsKvO"
    >
      <div
        className="sc-fFeiMQ ejkoIZ"
      >
        <div
          className="sc-bkkeKt bUsYpp"
        >
          Github Activity
        </div>
        <div
          className="sc-bkkeKt bUsYpp"
        >
          <span
            className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorSuccess css-15u8j6q-MuiCircularProgress-root"
            role="progressbar"
            style={
              Object {
                "height": 40,
                "width": 40,
              }
            }
          >
            <svg
              className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx={44}
                cy={44}
                fill="none"
                r={20.2}
                strokeWidth={3.6}
                style={Object {}}
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
